staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -30,6 +30,8 @@ const (
 	NamespaceAll string = ""
 	// NamespaceNodeLease is the namespace where we place node lease objects (used for node heartbeats)
 	NamespaceNodeLease string = "kube-node-lease"
+	// NamespaceAPIServerLease is the namespace where we place kube-apiserver lease objects (used for apiserver identity)
+	NamespaceAPIServerLease string = "kube-apiserver-lease"
```

Comment:
Why do we need an entire namespace for this? Is kube-system not good enough?

Reply:
There could be non-"apisever identity" leases created in kube-system, and we could accidentally garbage collect them.

Reply:
Actually we can use labels to avoid that. 

The separate-namespace idea is inherited from the node lease. In the [node lease KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0009-node-heartbeat.md) the reason was: 
> we considered using kube-system namespace but decided that it's already too overloaded

Reply:
I thought leases had specific names? It's really all leases in the namespace? I'd really expect multiple components to use leader election in the same namespace.

Reply:
> I thought leases had specific names?

Leader-election leases do (one name for one component), heartbeat leases don't (different name for each replica). 

> It's really all leases in the namespace?

Yes it is :( 

I agree leader-election leases for system components should live in the same namespace (kube-system). I prefer having heartbeat leases living in their own namespaces (kube-apiserver-lease) because of the clarity. Otherwise people&code need to label select leases in kube-system to distinguish these two types of leases. More importantly it allows the replicas to scale. 

Different kinds of leases:

```
$ kubectl get lease --all-namespaces
NAMESPACE              NAME                      HOLDER           AGE
kube-node-lease        node-a                    node-a
kube-node-lease        node-b                    node-b
kube-node-lease        node-c                    node-c
... (can be hundreds/thousands of them as the cluster scales)
kube-node-lease        node-zz                   node-zz

# these components use Endpoints for leader election by default, but can be configured to use Leases
kube-system            kube-controller-manager   hostname-a-UUID-a
kube-system            kube-scheduler            hostname-b-UUID-b

# **This PR adds**: (usually three objects; frequent apiserver restarts can add more)
kube-apiserver-lease   UUID-c                    UUID-c
kube-apiserver-lease   UUID-d                    UUID-d
kube-apiserver-lease   UUID-e                    UUID-e
kube-apiserver-lease   UUID-f                    UUID-f (stale record)

# user-installed controller (kubernetes-csi/external-attacher as an example)
NAMESPACE              NAME                               HOLDER           AGE
ns-that-runs-the-pods  external-attacher-leader-csi/dummy hostname-c
```

Reply:
discussed offline. We agreed to 1. put the leases in kube-system 2. name them "kube-apiserver" + UUID 3. label them so they can be selected by garbage collection.